cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -580,12 +582,30 @@ public SearchResults apply(MetadataDataset input) throws Exception {
   }
 
   @Override
-  public void rebuildIndexes(MetadataScope scope) throws DatasetManagementException, IOException {
+  public void rebuildIndexes(MetadataScope scope,
+                             RetryStrategy retryStrategy) {
     byte[] row = null;
-    while ((row = rebuildIndex(row, scope)) != null) {
+    while ((row = rebuildIndexesWithRetries(scope, row, retryStrategy)) != null) {
       LOG.debug("Completed a batch for rebuilding {} metadata indexes.", scope);
     }
+  }
 
+  private byte[] rebuildIndexesWithRetries(final MetadataScope scope,
```

Comment:
We have to stop wrapping checked exceptions into RuntimeExceptions everywhere.  

This should throw DatasetManagementException and IOException, like the rebuildIndexes() that calls it. And it should only throw a RuntimeException to wrap other exceptions (which would rarely happen).